DESIGN.md

```diff
@@ -0,0 +1,123 @@
+
+For Authorization, a simple form of role based access control will be implemented: Each role will have its own CA. All CAs will be trusted by the server. Upon handling a remote procedure call, the server will first check the client certificate's certificate authority and compare it to its internal mapping of certificate authority to privilege level. The following roles will be implemented:
+
+- **Read only:** this role can get job output, but not create or delete jobs. This is useful for automated tooling such as status indicators and job notifiers.
+- **Read/Write access:** this role can get job output as well as create and delete jobs. This role is useful for administrators and any kind of deployment automation that would leverage this service.
+
+![visual representation of the multiple client CA model](doc/client_cas.png)
+
+## API
+### Endpoints
+The API exposed by the RPC service will contain the following endpoints. Any errors mentioned refer to the [gRPC status codes](https://developers.google.com/maps-booking/reference/grpc-api/status_codes).
+
+#### **CreateJob**
+Receives a Command and returns a Job. Can return the following errors:
+
+- PERMISSION_DENIED: the client certificate is not signed by a CA that permits the read/write role
+- ABORTED: the command failed to launch
+
+#### **DeleteJob**
+Receives a JID and returns a Job. Can return the following errors:
+
+- PERMISSION_DENIED: the client certificate is not signed by a CA that permits the read/write role
+- NOT_FOUND: the JID was not found in the process cache
+
+After DeleteJob is run, the Job will be completely deallocated from the RPC system.
+
+#### **ListJobs**
+Returns a JobList. 
+
+#### **GetJob**
+Receives a JID and returns a Job (including updated job output and execution time).
+
+### Datatypes
+- **JID**
+    Single integer representing the ID of the Job as tracked by the RPC subsystem. Custom Job IDs were added to the implementation because it could not be assured that every Job would have a unique PID for the underlying process. 
+- **Command**
+    String that holds a shell command to be executed on the host system
+- **Status**
+    Enumeration that can hold the following values:
+    - Created
+    - In Process
+    - Stopped
+    - Exited
+    - Error
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.   
```

Comment:
If I read this correctly, the intention is that the server will execute a shell with the provided command?

So my process list would be something like:
server
bash -c "command"
"command"

What happens if I pass a command that contains quotes? IE something like `rpc_client --op CreateJob --arg "echo \"Hello World!\""`

Reply:
In your example you have escaped the quotations. It is most likely that your shell will parse two arguments: 'echo' and '"Hello World!"'

Consider the following shell script: 
```sh
 #!/bin/sh 
 for i; do 
    echo $i 
 done
```

Here is it's behavior on my system:
```sh
/tmp/test ⌚ 10:45:11
$ ./test.sh "\"has quotes\" more" second    
"has quotes" more
second
```

As you can see the shell has unescaped the inner quotations, but it is still part of the same argument as "more".
I have verified that this behavior is the same with zsh and bash. 

In conclusion, rpc_client will run the CreateJob routine with the argument `echo "Hello World!` on the host system.